File: frontend/src/services/api/prices.ts
/**
 * Price API functions for fetching market data
 */
import { apiClient } from './client'
import type { PricePoint, PriceHistory } from '@/types/price'
import { parseApiError } from '@/utils/priceErrors'

/**
 * Fetch current price for a ticker
 */
export async function getCurrentPrice(ticker: string): Promise<PricePoint> {
  // Backend returns flat structure, need to transform to nested PricePoint
  const response = await apiClient.get<{
    ticker: string
    price: string
    currency: string
    timestamp: string
    source: string
    is_stale: boolean
  }>(`/prices/${ticker}`)

  const data = response.data

  // Transform flat API response to nested PricePoint structure
  return {
    ticker: { symbol: data.ticker },
    price: {
      amount: parseFloat(data.price),
      currency: data.currency,
    },
    timestamp: data.timestamp,
    source: data.source as 'alpha_vantage' | 'cache' | 'database',
    interval: 'real-time', // Current price endpoint always returns real-time
  }
}

/**
 * Check if historical price data exists for a ticker at a specific date
 * Used by backtest mode to verify data availability
 */
export async function checkHistoricalPrice(
  ticker: string,
  date: string
): Promise<{ available: boolean; closest_date?: string }> {
  const response = await apiClient.get<{
    available: boolean
    closest_date: string | null
  }>(`/prices/${ticker}/check`, {
    params: { date },
  })

  return {
    available: response.data.available,
    closest_date: response.data.closest_date || undefined,
  }
}

/**
 * Get historical price for a ticker at a specific date
 * Returns the closest available price if exact date not available
 */
export async function getHistoricalPrice(
  ticker: string,
  date: string
): Promise<PricePoint> {
  // Use the /history endpoint with a 1-day range around the target date
  const response = await apiClient.get<{
    ticker: string
    prices: Array<{
      ticker: string
      price: string
      currency: string
      timestamp: string
      source: string
      interval: string
    }>
    start: string
    end: string
    interval: string
    count: number
  }>(`/prices/${ticker}/history`, {
    params: {
      start: date,
      end: date,
      interval: '1day',
    },
  })

  if (!response.data.prices || response.data.prices.length === 0) {
    throw new Error(`No price data available for ${ticker} at ${date}`)
  }

  // Return first (and should be only) price point
  const priceData = response.data.prices[0]
  return {
    ticker: { symbol: priceData.ticker },
    price: {
      amount: parseFloat(priceData.price),
      currency: priceData.currency,
    },
    timestamp: priceData.timestamp,
    source: priceData.source as 'alpha_vantage' | 'cache' | 'database',
    interval: priceData.interval as
      | '1day'
      | 'real-time'
      | '1hour'
      | '5min'
      | '1min',
  }
}


/**
 * Batch fetch prices for multiple tickers
 * Uses the /prices/batch endpoint for efficient single-request fetching
 */
export async function getBatchPrices(
  tickers: string[]
): Promise<Map<string, PricePoint>> {
  // Return empty map if no tickers provided
  if (tickers.length === 0) {
    return new Map()
  }

  // Call batch endpoint with comma-separated tickers
  const response = await apiClient.get<{
    prices: Record<
      string,
      {
        ticker: string
        price: string
        currency: string
        timestamp: string
        source: string
        is_stale: boolean
      }
    >
    requested: number
    returned: number
  }>('/prices/batch', {
    params: { tickers: tickers.join(',') },
  })

  // Convert backend response format to Map<string, PricePoint>
  const priceMap = new Map<string, PricePoint>()
  for (const [ticker, priceData] of Object.entries(response.data.prices)) {
    // Validate source value or default to 'database'
    const validSources: Array<'alpha_vantage' | 'cache' | 'database'> = [
      'alpha_vantage',
      'cache',
      'database',
    ]
    const source = validSources.includes(
      priceData.source as 'alpha_vantage' | 'cache' | 'database'
    )
      ? (priceData.source as 'alpha_vantage' | 'cache' | 'database')
      : 'database'

    priceMap.set(ticker, {
      ticker: { symbol: priceData.ticker },
      price: {
        amount: parseFloat(priceData.price),
        currency: priceData.currency,
      },
      timestamp: priceData.timestamp,
      source,
      interval: 'real-time',
    })
  }

  return priceMap
}

/**
 * Fetch price history for a ticker within a date range
 * Backend returns prices as strings, we parse them to numbers
 */
export async function getPriceHistory(
  ticker: string,
  startDate: string,
  endDate: string
): Promise<PriceHistory> {
  try {
    // Backend returns prices as strings, need to parse to numbers
    const response = await apiClient.get<{
      ticker: string
      prices: Array<{
        ticker: string
        price: string // Backend sends string
        currency: string
        timestamp: string
        source: string
        interval: string
      }>
    }>(`/prices/${ticker}/history`, {
      params: { start: startDate, end: endDate },
    })

    // Convert backend response to PriceHistory with number prices
    const priceHistory: PriceHistory = {
      ticker: response.data.ticker,
      prices: response.data.prices.map((point) => ({
        ticker: { symbol: point.ticker },
        price: {
          amount: parseFloat(point.price), // Parse string to number
          currency: point.currency,
        },
        timestamp: point.timestamp,
        source: point.source as 'alpha_vantage' | 'cache' | 'database',
        interval: point.interval as
          | '1day'
          | 'real-time'
          | '1hour'
          | '5min'
          | '1min',
      })),
      source: response.data.prices[0]?.source || 'unknown',
      cached: response.data.prices[0]?.source === 'cache',
    }

    return priceHistory
  } catch (error) {
    // Parse error into ApiError type
    const apiError = parseApiError(error, ticker)

    // In development mode, show mock data with warning banner
    if (import.meta.env.DEV) {
      console.warn('[DEV] API error, using mock data:', apiError.message)
      return {
        ...generateMockPriceHistory(ticker, startDate, endDate),
        error: apiError, // Include error in response
      }
    }

    // In production, throw structured error for component to handle
    throw apiError
  }
}

/**
 * Generate mock price history data for development
 * This will be removed once the backend API is available
 */
function generateMockPriceHistory(
  ticker: string,
  startDate: string,
  endDate: string
): PriceHistory {
  const start = new Date(startDate)
  const end = new Date(endDate)
  const days = Math.ceil(
    (end.getTime() - start.getTime()) / (1000 * 60 * 60 * 24)
  )

  const prices: PricePoint[] = []
  let basePrice = 150 + Math.random() * 100 // Random starting price between 150-250

  for (let i = 0; i <= days; i++) {
    const date = new Date(start)
    date.setDate(start.getDate() + i)

    // Random walk with some volatility
    const change = (Math.random() - 0.5) * 5
    basePrice = Math.max(50, basePrice + change) // Don't go below 50

    prices.push({
      ticker: { symbol: ticker },
      price: {
        amount: Number(basePrice.toFixed(2)),
        currency: 'USD',
      },
      timestamp: date.toISOString(),
      source: 'cache',
      interval: '1day',
    })
  }

  return {
    ticker,
    prices,
    source: 'mock',
    cached: false,
  }
}

export const pricesApi = {
  getCurrentPrice,
  getBatchPrices,
  getPriceHistory,
}

File: backend/src/zebu/adapters/inbound/api/prices.py
"""Price API routes.

Provides REST endpoints for price data operations:
- Get current price for a ticker
- Get historical price data for a ticker
- Get supported tickers
"""

from datetime import UTC, datetime, timedelta
from typing import Annotated

import structlog
from fastapi import APIRouter, HTTPException, Query, status
from pydantic import BaseModel, Field

from zebu.adapters.inbound.api.dependencies import MarketDataDep
from zebu.application.exceptions import (
    MarketDataUnavailableError,
    TickerNotFoundError,
)
from zebu.domain.value_objects.ticker import Ticker

logger = structlog.get_logger(__name__)

router = APIRouter(prefix="/prices", tags=["prices"])


# Response Models


class PricePointResponse(BaseModel):
    """Response model for a single price point."""

    ticker: str = Field(..., description="Stock ticker symbol")
    price: str = Field(..., description="Price in decimal format")
    currency: str = Field(..., description="ISO 4217 currency code")
    timestamp: datetime = Field(..., description="When price was observed (UTC)")
    source: str = Field(..., description="Data source (alpha_vantage, cache, database)")
    interval: str = Field(..., description="Price interval type")


class PriceHistoryResponse(BaseModel):
    """Response model for price history query."""

    ticker: str = Field(..., description="Stock ticker symbol")
    prices: list[PricePointResponse] = Field(..., description="List of price points")
    start: datetime = Field(..., description="Start of time range (UTC)")
    end: datetime = Field(..., description="End of time range (UTC)")
    interval: str = Field(..., description="Price interval type")
    count: int = Field(..., description="Number of price points returned")


class CurrentPriceResponse(BaseModel):
    """Response model for current price query."""

    ticker: str = Field(..., description="Stock ticker symbol")
    price: str = Field(..., description="Current price in decimal format")
    currency: str = Field(..., description="ISO 4217 currency code")
    timestamp: datetime = Field(..., description="When price was observed (UTC)")
    source: str = Field(..., description="Data source")
    is_stale: bool = Field(..., description="Whether price data is stale")


class BatchPriceItem(BaseModel):
    """Single price item in batch response."""

    ticker: str = Field(..., description="Stock ticker symbol")
    price: str = Field(..., description="Current price in decimal format")
    currency: str = Field(..., description="ISO 4217 currency code")
    timestamp: datetime = Field(..., description="When price was observed (UTC)")
    source: str = Field(..., description="Data source (alpha_vantage, cache, database)")
    is_stale: bool = Field(..., description="Whether price data is stale")


class BatchPriceResponse(BaseModel):
    """Response model for batch price query."""

    prices: dict[str, BatchPriceItem] = Field(
        ..., description="Mapping of ticker symbols to price data"
    )
    requested: int = Field(..., description="Number of tickers requested")
    returned: int = Field(..., description="Number of prices returned")


class SupportedTickersResponse(BaseModel):
    """Response model for supported tickers query."""

    tickers: list[str] = Field(..., description="List of supported ticker symbols")
    count: int = Field(..., description="Number of tickers")


class CheckHistoricalDataResponse(BaseModel):
    """Response model for historical data availability check."""

    available: bool = Field(..., description="Whether historical data exists")
    closest_date: datetime | None = Field(
        None, description="Closest available date if data exists"
    )


class FetchHistoricalDataRequest(BaseModel):
    """Request model for fetching historical data."""

    ticker: str = Field(..., min_length=1, max_length=5)
    start: datetime = Field(..., description="Start date (UTC)")
    end: datetime = Field(..., description="End date (UTC)")


class FetchHistoricalDataResponse(BaseModel):
    """Response model for historical data fetch."""

    ticker: str = Field(..., description="Stock ticker symbol")
    fetched: int = Field(..., description="Number of price points fetched")
    start: datetime = Field(..., description="Start of time range (UTC)")
    end: datetime = Field(..., description="End of time range (UTC)")


# Route Handlers


@router.get(
    "/batch",
    response_model=BatchPriceResponse,
    summary="Get current prices for multiple tickers",
    description="Fetches current prices for multiple tickers in a single batch request",
)
async def get_batch_prices(
    tickers: Annotated[
        str,
        Query(
            description=(
                "Comma-separated list of ticker symbols (e.g., 'AAPL,MSFT,GOOGL')"
            )
        ),
    ],
    market_data: MarketDataDep,
) -> BatchPriceResponse:
    """Get current prices for multiple tickers in batch.

    This endpoint optimizes price fetching by:
    - Checking cache for all tickers first
    - Only fetching uncached tickers from API
    - Returning partial results if some tickers fail

    Args:
        tickers: Comma-separated ticker symbols (e.g., "AAPL,MSFT,GOOGL")
        market_data: Market data port implementation (injected)

    Returns:
        BatchPriceResponse with prices for available tickers

    Example:
        GET /api/v1/prices/batch?tickers=AAPL,MSFT,GOOGL
    """
    # Parse comma-separated tickers
    ticker_list = [t.strip().upper() for t in tickers.split(",") if t.strip()]

    if not ticker_list:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="At least one ticker symbol is required",
        )

    # Convert to Ticker objects
    ticker_objs = [Ticker(t) for t in ticker_list]

    # Fetch batch prices
    prices_dict = await market_data.get_batch_prices(ticker_objs)

    # Convert to response format
    prices_response: dict[str, BatchPriceItem] = {}
    for ticker_obj, price_point in prices_dict.items():
        prices_response[ticker_obj.symbol] = BatchPriceItem(
            ticker=price_point.ticker.symbol,
            price=str(price_point.price.amount),
            currency=price_point.price.currency,
            timestamp=price_point.timestamp,
            source=price_point.source,
            is_stale=price_point.is_stale(max_age=timedelta(hours=1)),
        )

    return BatchPriceResponse(
        prices=prices_response,
        requested=len(ticker_list),
        returned=len(prices_response),
    )


@router.get(
    "/{ticker}",
    response_model=CurrentPriceResponse,
    summary="Get current price for a ticker",
    description="Fetches the most recent available price for a stock ticker",
)
async def get_current_price(
    ticker: str,
    market_data: MarketDataDep,
) -> CurrentPriceResponse:
    """Get current price for a ticker.

    Args:
        ticker: Stock ticker symbol (e.g., "AAPL")
        market_data: Market data port implementation (injected)

    Returns:
        CurrentPriceResponse with current price data

    Raises:
        HTTPException: 404 if ticker not found, 503 if market data unavailable
    """
    try:
        # Parse ticker
        ticker_obj = Ticker(ticker.upper())

        # Get current price
        price_point = await market_data.get_current_price(ticker_obj)

        # Convert to response model
        return CurrentPriceResponse(
            ticker=price_point.ticker.symbol,
            price=str(price_point.price.amount),
            currency=price_point.price.currency,
            timestamp=price_point.timestamp,
            source=price_point.source,
            is_stale=price_point.is_stale(max_age=timedelta(hours=1)),
        )

    except TickerNotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Ticker not found: {ticker}",
        ) from e

    except MarketDataUnavailableError as e:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=str(e),
        ) from e


@router.get(
    "/{ticker}/history",
    response_model=PriceHistoryResponse,
    summary="Get historical price data",
    description="Fetches historical price data for a ticker over a time range",
)
async def get_price_history(
    ticker: str,
    market_data: MarketDataDep,
    start: Annotated[
        str, Query(description="Start of time range (YYYY-MM-DD or ISO datetime)")
    ],
    end: Annotated[
        str, Query(description="End of time range (YYYY-MM-DD or ISO datetime)")
    ],
    interval: Annotated[
        str,
        Query(description="Price interval (1min, 5min, 1hour, 1day)"),
    ] = "1day",
) -> PriceHistoryResponse:
    """Get historical price data for a ticker.

    Args:
        ticker: Stock ticker symbol (e.g., "AAPL")
        market_data: Market data port implementation (injected)
        start: Start of time range (UTC)
        end: End of time range (UTC)
        interval: Price interval type (default: "1day")

    Returns:
        PriceHistoryResponse with list of price points

    Raises:
        HTTPException: 400 if invalid parameters, 404 if ticker not found,
                      503 if market data unavailable
    """
    try:
        # Parse ticker
        ticker_obj = Ticker(ticker.upper())

        # Parse dates - ensure timezone-aware datetimes (UTC)
        # Supports both "2026-01-12" (naive) and "2026-01-12T00:00:00Z" (aware) formats

        start_parsed = datetime.fromisoformat(start.replace("Z", "+00:00"))
        end_parsed = datetime.fromisoformat(end.replace("Z", "+00:00"))

        # If naive (no timezone), assume UTC
        start_dt = (
            start_parsed if start_parsed.tzinfo else start_parsed.replace(tzinfo=UTC)
        )
        end_dt = end_parsed if end_parsed.tzinfo else end_parsed.replace(tzinfo=UTC)

        # Bind ticker context for all logs in this request handler
        log = logger.bind(ticker=ticker, interval=interval)

        log.info(
            "Price history API request",
            start=start_dt.isoformat(),
            end=end_dt.isoformat(),
        )

        # Adjust end date to include full day if it's exactly midnight
        # When frontend sends "2026-01-17", it gets parsed as "2026-01-17T00:00:00"
        # but we want to include all data points on that day (up to 23:59:59.999999)
        adjusted_end = end_dt
        if end_dt.time() == end_dt.min.time():  # Check if time is 00:00:00
            from datetime import timedelta

            adjusted_end = end_dt + timedelta(days=1, microseconds=-1)

            log.debug(
                "Adjusted end date for midnight boundary",
                original_end=end_dt.isoformat(),
                adjusted_end=adjusted_end.isoformat(),
            )

        # Get price history
        history = await market_data.get_price_history(
            ticker_obj, start_dt, adjusted_end, interval
        )

        # Convert to response model
        prices = [
            PricePointResponse(
                ticker=p.ticker.symbol,
                price=str(p.price.amount),
                currency=p.price.currency,
                timestamp=p.timestamp,
                source=p.source,
                interval=p.interval,
            )
            for p in history
        ]

        # Log response metadata
        log.info(
            "Price history API response",
            count=len(prices),
            status="success",
        )

        # Warn if response is empty
        if len(prices) == 0:
            log.warning(
                "Price history returned empty",
                start=start_dt.isoformat(),
                end=adjusted_end.isoformat(),
            )

        return PriceHistoryResponse(
            ticker=ticker_obj.symbol,
            prices=prices,
            start=start_dt,
            end=adjusted_end,
            interval=interval,
            count=len(prices),
        )

    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        ) from e

    except TickerNotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Ticker not found: {ticker}",
        ) from e

    except MarketDataUnavailableError as e:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=str(e),
        ) from e


@router.get(
    "/",
    response_model=SupportedTickersResponse,
    summary="Get supported tickers",
    description="Returns list of all tickers we have price data for",
)
async def get_supported_tickers(
    market_data: MarketDataDep,
) -> SupportedTickersResponse:
    """Get list of supported tickers.

    Args:
        market_data: Market data port implementation (injected)

    Returns:
        SupportedTickersResponse with list of ticker symbols

    Raises:
        HTTPException: 503 if market data unavailable
    """
    try:
        tickers = await market_data.get_supported_tickers()

        return SupportedTickersResponse(
            tickers=[t.symbol for t in tickers],
            count=len(tickers),
        )

    except MarketDataUnavailableError as e:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=str(e),
        ) from e


@router.get(
    "/{ticker}/check",
    response_model=CheckHistoricalDataResponse,
    summary="Check historical data availability",
    description="Check if historical price data exists for a ticker at a specific date",
)
async def check_historical_data(
    ticker: str,
    date: Annotated[datetime, Query(description="Date to check (UTC)")],
    market_data: MarketDataDep,
) -> CheckHistoricalDataResponse:
    """Check if historical data exists for a ticker on a specific date.

    This endpoint is used by the frontend to determine if historical price data
    is available before executing a backtest trade.

    Args:
        ticker: Stock ticker symbol (e.g., "AAPL")
        date: Date to check for data availability (UTC)
        market_data: Market data port implementation (injected)

    Returns:
        CheckHistoricalDataResponse with availability status

    Raises:
        HTTPException: 400 if invalid ticker format, 503 if service unavailable
    """
    try:
        # Parse ticker
        ticker_obj = Ticker(ticker.upper())

        # Try to get price at the specified date
        price = await market_data.get_price_at(ticker_obj, date)

        return CheckHistoricalDataResponse(
            available=True,
            closest_date=price.timestamp,
        )

    except TickerNotFoundError:
        # Ticker not found - return available=False instead of error
        # This allows frontend to handle gracefully
        return CheckHistoricalDataResponse(
            available=False,
            closest_date=None,
        )

    except MarketDataUnavailableError:
        # No data available - return False instead of error
        return CheckHistoricalDataResponse(
            available=False,
            closest_date=None,
        )


@router.post(
    "/fetch-historical",
    response_model=FetchHistoricalDataResponse,
    summary="Fetch historical data",
    description="Fetch and store historical price data for a ticker",
)
async def fetch_historical_data(
    request: FetchHistoricalDataRequest,
    market_data: MarketDataDep,
) -> FetchHistoricalDataResponse:
    """Fetch historical data for a ticker and date range.

    This endpoint is called by the frontend when backtest mode detects missing
    historical data. It fetches data from Alpha Vantage and stores it in the
    database for future use.

    Args:
        request: Request with ticker and date range
        market_data: Market data port implementation (injected)

    Returns:
        FetchHistoricalDataResponse with fetch results

    Raises:
        HTTPException: 400 if invalid parameters, 404 if ticker not found,
                      503 if market data unavailable
    """
    try:
        # Parse ticker
        ticker_obj = Ticker(request.ticker.upper())

        # Fetch price history from API (will auto-store in database)
        history = await market_data.get_price_history(
            ticker_obj,
            start=request.start,
            end=request.end,
            interval="1day",
        )

        return FetchHistoricalDataResponse(
            ticker=ticker_obj.symbol,
            fetched=len(history),
            start=request.start,
            end=request.end,
        )

    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        ) from e

    except TickerNotFoundError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Ticker not found: {request.ticker}",
        ) from e

    except MarketDataUnavailableError as e:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=str(e),
        ) from e


@router.get(
    "/debug/cache/{ticker}",
    summary="Inspect price cache (Debug only)",
    description="Development endpoint to inspect cached price data for debugging",
    include_in_schema=True,
)
async def inspect_price_cache(
    ticker: str,
    market_data: MarketDataDep,
) -> dict[str, object]:
    """Inspect cached price data for debugging.

    **Development only** - shows what data exists in cache for a ticker.
    This endpoint helps diagnose data completeness issues by showing:
    - Total cached points for the ticker
    - Date range of cached data
    - List of all cached dates
    - Potential gaps in the data

    Args:
        ticker: Stock ticker symbol (e.g., "AAPL")
        market_data: Market data port implementation (injected)

    Returns:
        Dictionary with cache inspection data
    """
    try:
        from zebu.adapters.outbound.market_data.alpha_vantage_adapter import (
            AlphaVantageAdapter,
        )

        ticker_obj = Ticker(ticker.upper())

        # Get all data for this ticker (last 100 days)
        end = datetime.now()
        start = end - timedelta(days=100)

        # Check if we have access to price repository
        # (only available on AlphaVantageAdapter implementation)
        if not isinstance(market_data, AlphaVantageAdapter):
            return {
                "ticker": ticker,
                "status": "error",
                "message": "Debug endpoint only works with AlphaVantageAdapter",
            }

        if not market_data.price_repository:
            return {
                "ticker": ticker,
                "status": "error",
                "message": "Price repository not configured",
            }

        history = await market_data.price_repository.get_price_history(
            ticker_obj, start, end, "1day"
        )

        if not history:
            return {
                "ticker": ticker,
                "status": "no_data",
                "message": "No cached data found for this ticker",
            }

        # Extract dates and find gaps
        dates = [p.timestamp.date() for p in history]
        date_strings = [d.isoformat() for d in dates]

        # Simple gap detection: find dates more than 7 days apart
        # (allowing for weekends/holidays)
        gaps = []
        for i in range(len(dates) - 1):
            days_apart = (dates[i + 1] - dates[i]).days
            # More than a week gap (accounting for weekends/holidays)
            if days_apart > 7:
                gaps.append(
                    {
                        "after": dates[i].isoformat(),
                        "before": dates[i + 1].isoformat(),
                        "days_gap": days_apart,
                    }
                )

        return {
            "ticker": ticker,
            "status": "ok",
            "total_points": len(history),
            "date_range": {
                "start": history[0].timestamp.isoformat(),
                "end": history[-1].timestamp.isoformat(),
                "span_days": (dates[-1] - dates[0]).days,
            },
            "dates": date_strings,
            "gaps_detected": gaps,
            "sources": list({p.source for p in history}),
        }

    except Exception as e:
        logger.error(
            "Error inspecting price cache",
            ticker=ticker,
            error=str(e),
            exc_info=True,
        )
        return {
            "ticker": ticker,
            "status": "error",
            "message": f"Error inspecting cache: {str(e)}",
        }
